controller/src/main/java/io/pravega/controller/server/Main.java

```diff
@@ -86,8 +88,30 @@ public static void main(String[] args) {
                     .restServerConfig(Optional.of(restServerConfig))
                     .build();
 
+            Thread.setDefaultUncaughtExceptionHandler((t, e) -> {
+                log.error("Thread {} with stackTrace {} failed with uncaught exception", t.getName(), t.getStackTrace(), e);              
+            });
+
             ControllerServiceMain controllerServiceMain = new ControllerServiceMain(serviceConfig);
             controllerServiceMain.startAsync();
+            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
+                memoryMXBean.setVerbose(true);
+                log.error("Shutdown hook memory usage dump: Heap memory usage: {}, non heap memory usage {}", memoryMXBean.getHeapMemoryUsage(),
+                        memoryMXBean.getNonHeapMemoryUsage());
+                
+                log.info("Controller service shutting down");
+                try {
+                    controllerServiceMain.stopAsync();
+                    controllerServiceMain.awaitTerminated();
+                } finally {
+                    if (Config.DUMP_STACK_ON_SHUTDOWN) {
+                        Thread.getAllStackTraces().forEach((key, value) ->
+                                log.info("Shutdown Hook Thread dump: Thread {} stackTrace: {} ", key.getName(), value));
```

Comment:
This might be a good idea, but let me try to understand the reasoning behind it. Are you assuming that the JVM is shutting down due to an error? The issue talks about a clean shutdown, so I'm trying to match what I'm seeing here to the issue description.

Reply:
Basically mesos, k8s and other package managers sometime shutdown instances if they cross resource limits. it may be good to know if when resource limits are exceeding, what was the snapshot of threads and memory footprint in general. 
but we dont always want to enable the flag.. only if it happens enough, we could ask customers to turn on the flag and have logs collected in such instances. 

By default in our system tests and longevity, having these dumped is a good idea..